src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -736,6 +736,11 @@ void PrepareFilterParameter()
 
 					HandleWithFragment(fromElement, with);
 				}
+
+				if (fromElement.Parent == null)
```

Comment:
Shouldn't it be done where the `fromElement` is created?

Reply:
yes, but there are cases where the item is added to the tree explicitly in FromElement.SetOrigin but there is an else branch with comment
`				// Otherwise, the destination node was implied by the FROM clause and the FROM clause processor
				// will automatically add it in the right place.
`

But at least for the implied join in an explicit join no code took care of adding the item to the tree.

I tried to identify the case and go to the previous else if branch (where the comment HHH-276 is located) but that caused one legacy unit test to fail.

As there is no clear responsibility for how the element should be added to the tree, I found it most easy to add it where I did. It's some kind of catch all cases where the item is not added yet which one can find good or bad. But as I said, there is no clear responsibility so I tended towards the feeling: it's a good thing.